### ISSUE TITLE
Mismatch In output of target= NEON and target=CL output for Vs 20.11 using tensorflow based network which is working well for Vs 19.05

### ISSUE THREAD

**Issue by abhajaswal** [ARM-software/ComputeLibrary#867]
**Output of 'strings libarm_compute.so | grep arm_compute_version':**

arm_compute_version=v20.11 Build options: {'embed_kernels': '1', 'os': 'tizen', 'utils': '1', 'opencl': '1', 'neon': '1', 'benchmark_tests': '0', 'validation_tests': '0', 'arch': 'armv7a', 'asserts': '1', 'debug': '1', 'openmp': '1', 'cppthreads': '1', 'Werror': '0', 'examples': '1'} Git hash=unknown

**Platform:**
armv7a

**Operating System:**
tizen

I am trying to run a tensorflow based network which is working fine for 19.05 for both targets CL and NEON.
Output of both targets match with the tensorflow ,(just minor precision variation as compared to refernce tensorflow)

But when is ported the same solution on 20.11.  The neon output  match with tensorflow output, but the target CL output is a mismatch.
I also tried layer wise output dump. The output of NEON and CL match uptil third Pooling layer (name - pool3 ) layer only 
The last batch normalization layer (name - subBN) output becomes incorrect for target CL . Both these layer i have marked bold in below graph log-

**Problem description:**

Please find the log for the network run:
-----------------------------------------------------------------------
./graph_model4_sign

Threads : 1
Target : CL
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : true
Tuner mode : Normal
Tuner file :
Fast math enabled? : false
Data path : /usr/share/cl-res/

INIT START
INIT is OVER
 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : NodeFusionMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : GroupedConvolutionMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : InPlaceOperationMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : DepthConcatSubTensorMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : SplitLayerSubTensorMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Running mutating pass : NodeExecutionMethodMutator

 [GRAPH][08-12-2020 05:57:12][INFO]  Instantiated conv1 Type: GEMMConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 1x64x96 Weights shape: 1x5x5x32 Output shape: 32x60x92 RELU

 [GRAPH][08-12-2020 05:57:13][INFO]  Instantiated pool1 Type: PoolingLayer Target: CL Data Type: F32 Input shape: 32x60x92 Output shape: 32x30x46 Pooling info: MAX

 [GRAPH][08-12-2020 05:57:13][INFO]  Instantiated conv2 Type: GEMMConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 32x30x46 Weights shape: 32x3x3x16 Output shape: 16x28x44 RELU

 [GRAPH][08-12-2020 05:57:13][INFO]  Instantiated pool2 Type: PoolingLayer Target: CL Data Type: F32 Input shape: 16x28x44 Output shape: 16x14x22 Pooling info: MAX

 [GRAPH][08-12-2020 05:57:14][INFO]  Instantiated conv3 Type: GEMMConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 16x14x22 Weights shape: 16x3x3x16 Output shape: 16x12x20

 [GRAPH][08-12-2020 05:57:14][INFO]  Instantiated pool3 Type: PoolingLayer Target: CL Data Type: F32 Input shape: 16x12x20 Output shape: 16 Pooling info: AVG

 [GRAPH][08-12-2020 05:57:14][INFO]  Instantiated Reshape Type: ReshapeLayer Target: CL Data Type: F32 Input shape: 16 Output shape: 16

 [GRAPH][08-12-2020 05:57:14][INFO]  Instantiated subBN Type: BatchNormalizationLayer Target: CL Data Type: F32 Shape: 16 Epsilon: 0  InPlace: 1

 [GRAPH][08-12-2020 05:57:15][INFO]  Instantiated l2norm Type: L2NormalizeLayer Target: CL Data Type: F32 Input shape: 16 Output shape: 16 Axis: 0 Epsilon: 1e-12

 [GRAPH][08-12-2020 05:57:15][INFO]  Instantiated dense0 Type: FullyConnectedLayer Target: CL Data Type: F32 Input shape: 16 Weights shape: 4x16 Output shape: 4

 [GRAPH][08-12-2020 05:57:15][INFO]  Instantiated prob Type: SoftmaxLayer Target: CL Data Type: F32 Input shape: 4 Output shape: 4

------------------------------------------------------------------------------------------------------------------------------------------

The weights for last batch norm (gamma, variance, beta, mean are below)

BETA weights are 
array([[ 1.420612  , -1.2331833 , -1.8151608 , -0.03307388,  1.8477985 ,
        -0.9415488 ,  1.668008  , -1.0041965 , -2.2218935 , -1.0505359 ,
        -1.1981492 ,  0.9194711 , -2.1054451 ,  0.26712406, -0.85119057,
        -1.9043738 ]], dtype=float32)

GAMMA weights are  - 
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
      dtype=float32)

MEAN weights are -
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
      dtype=float32)

Variance weights are - 
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
      dtype=float32)

Epsilon value is 0.000f

    If I use these BN wieghts and use formula   

     X_norm = (X - mu) / np.sqrt(var + epsilon)
    out = gamma * X_norm + beta

 the result is of batch norm should be straight forward just addition of beta values and pooling layer 3 output

Output of Pooling layer 3 i,e input to last batch normalization  for both CL and Neon target is 
-1.407413 
1.227553 
1.819020 
0.064929 
-1.815518 
0.885946 
-1.635981 
1.019594 
2.238474 
1.058782 
1.208596 
-0.913676 
2.111688 
-0.287055 
0.845316 
1.916249 
---------------------------

output of Last  batch normliztion layer NEON(correct) is

0.013199
-0.00563
0.003859
0.031855
0.03228
-0.0556
0.032027
0.015398
0.01658
0.008246
0.010447
0.005795
0.006243
-0.01993
-0.00588
0.011875

And Last batch normalization layer output on acl v20.11  target  CL(Incorrect) is 
1.749870 
-1.999996 
-1.999477 
1.999998 
-1.815518 
0.885946 
-1.635981 
1.019594 
2.238474 
1.058782 
1.208596 
-0.913676 
2.111688 
-0.287055 
0.845316 
1.916249 

Let me know what is the issue with the same.
I have dumped each layer wise , output tensor using the graph node name and compared them for both targets.
Mismatch begin after pool3 layer

**Comment by abhajaswal** [comment]
Dear @morgolock were you able to check this at your end? Is there any workaround for same available so that I could proceed?

**Comment by morgolock** [comment]
Hi @abhajaswal 

I wrote a test but I'm having problems to reproduce the issue.

Could you please provide a standalone test reproducing the problem? You could modify the test below.

```
class NEONCLBatchNormExample : public Example
{
public:
    bool do_setup(int, char **) override
    {
        /** [NEON / OpenCL Interop] */

        CLScheduler::get().default_init();

        const std::vector<float> input_data = {
                -1.407413f,1.227553f, 1.819020f,0.064929f,-1.815518f,
                0.885946f,-1.635981f,1.019594f,2.238474f,1.058782f,
                1.208596f,-0.913676f,2.111688f,-0.287055f,0.845316f,
                1.916249f
        };

        const std::vector<float> beta_data = {
             1.420612f , -1.2331833f , -1.8151608f , -0.03307388f, 1.8477985f ,
            -0.9415488f , 1.668008f , -1.0041965f , -2.2218935f , -1.0505359f ,
            -1.1981492f , 0.9194711f , -2.1054451f , 0.26712406f, -0.85119057f,
            -1.9043738f };

        const auto input_size = input_data.size();
        const TensorInfo input_info(TensorInfo(TensorShape(input_size,1,16), Format::F32));

        src.allocator()->init(input_info);
        dst.allocator()->init(input_info);
        var.allocator()->init(TensorInfo(TensorShape(16),Format::F32));
        mean.allocator()->init(*var.info());
        beta.allocator()->init(*mean.info());
        gamma.allocator()->init(*mean.info());
 
        ne_bn.configure(&src, &dst, &mean, &var, &beta, &gamma, 1e-12/*, ActivationLayerInfo(ActivationLayerInfo::ActivationFunction::RELU)*/);

        // Allocate all the images
        src.allocator()->allocate();
        dst.allocator()->allocate();
        var.allocator()->allocate();
        mean.allocator()->allocate();
        beta.allocator()->allocate();
        gamma.allocator()->allocate();

        Window input_window;
        input_window.use_tensor_dimensions(src.info()->tensor_shape());
        // Create an iterator:
        Iterator input_it(&src, input_window);
        // Except it works for an arbitrary number of dimensions
        execute_window_loop(input_window, [&](const Coordinates & id)
        {
            *reinterpret_cast<float *>(input_it.ptr()) = input_data[id.x()];
        },
        input_it);

        Window mean_window;
        mean_window.use_tensor_dimensions(mean.info()->tensor_shape());
        Iterator mean_it(&mean, mean_window);
        Iterator var_it(&var, mean_window);
        Iterator gamma_it(&gamma, mean_window);
        Iterator beta_it(&beta, mean_window);
 
        // Except it works for an arbitrary number of dimensions
        execute_window_loop(mean_window, [&](const Coordinates &id)
        {
            *reinterpret_cast<float *>(mean_it.ptr()) = 0.f;
            *reinterpret_cast<float *>(var_it.ptr()) = 1.f;
            *reinterpret_cast<float *>(gamma_it.ptr()) = 1.f;
            *reinterpret_cast<float *>(beta_it.ptr()) = beta_data[id.x()]; 
        },
        mean_it,var_it, gamma_it,beta_it);
        return true;
    }
    void do_run() override
    {
        ne_bn.run();
        // Make sure all the OpenCL jobs are done executing:
        CLScheduler::get().sync();
    }
    void do_teardown() override
    {
        // Save the result to file:
        std::cout << "input" << std::endl;
        src.print(std::cout);
        std::cout << "beta" << std::endl;
        beta.print(std::cout);
        std::cout << "dst tensor" << std::endl;
        dst.print(std::cout);
    }
private:
    Tensor       src{}, dst{},mean{},var{},beta{},gamma{};
    CLBatchNormalizationLayer cl_bn{};
    NEBatchNormalizationLayer ne_bn{};
};
```

**Comment by abhajaswal** [comment]
Hi Morgolock ,

I do not have a standalone test app which I could share, due to security policies to open it up on portal.
Can there be any issue due to mutator in any of the graph pipeline?

**Comment by morgolock** [comment]
Hi @abhajaswal 

1. I cannot reproduce. In order to help I need a way to reproduce. You could make some changes to the code I've provided to reproduce the problem.
2. Another thing you could try is updating to v21.02.
3. Can you share the code setting up the model using the graph api?



>Can there be any issue due to mutator in any of the graph pipeline?

@GeorgeARM  could you please let us know what are your thoughts about the mutator?

**Comment by morgolock** [comment]
Closing due to inactivity, please reopen if support is still required
